In [84]:
import numpy as np, pandas as pd, matplotlib.pyplot as plt
from datetime import time, datetime

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
!pip install dash
!pip install jupyter_dash

In [85]:
import dash
from jupyter_dash import JupyterDash
from dash import html, dcc
from dash.dependencies import Input,Output,State

In [86]:
import plotly.express as px
import plotly.graph_objects as go

In [87]:
data= pd.read_excel('/content/drive/MyDrive/NewExcellData.xlsx')

In [88]:
df= data.copy()

In [89]:
df.head()

Orderdate  salesValue  ItemsQty  DeliverysalesValue  \
0 2022-02-24 16:23:54       122.5      4.25                 0.0   
1 2022-01-17 21:13:47        35.0      1.00                 0.0   
2 2022-03-11 20:54:27       157.5      7.00                 0.0   
3 2022-02-02 16:57:59        85.0      2.00                 0.0   
4 2022-02-03 17:47:44       100.0      1.00                 0.0   

   DeliveryitemsQty  OfferSales  TimePerorder  
0               0.0         0.0           5.0  
1               0.0         0.0           5.0  
2               0.0         0.0           5.0  
3               0.0         0.0           5.0  
4               0.0         0.0           5.0

In [90]:
df.shape

(829494, 7)

In [91]:
df['Orderdate'].min(), df['Orderdate'].max()

(Timestamp('2022-01-01 03:07:52'), Timestamp('2022-05-22 16:59:15'))

The data covers from Jan to May 2022

In [92]:
df.duplicated().value_counts()

False    826296
True       3198
dtype: int64

In [93]:
df.isna().sum()

Orderdate             0
salesValue            0
ItemsQty              0
DeliverysalesValue    0
DeliveryitemsQty      0
OfferSales            0
TimePerorder          0
dtype: int64

In [55]:
# df['Date'], df['Time'] = pd.to_datetime(df['Orderdate']).dt.date, pd.to_datetime(df['Orderdate']).dt.time

In [94]:
df['Date'], df['Time'] = df['Orderdate'].dt.date, df['Orderdate'].dt.time

In [95]:
df.head()

Orderdate  salesValue  ItemsQty  DeliverysalesValue  \
0 2022-02-24 16:23:54       122.5      4.25                 0.0   
1 2022-01-17 21:13:47        35.0      1.00                 0.0   
2 2022-03-11 20:54:27       157.5      7.00                 0.0   
3 2022-02-02 16:57:59        85.0      2.00                 0.0   
4 2022-02-03 17:47:44       100.0      1.00                 0.0   

   DeliveryitemsQty  OfferSales  TimePerorder        Date      Time  
0               0.0         0.0           5.0  2022-02-24  16:23:54  
1               0.0         0.0           5.0  2022-01-17  21:13:47  
2               0.0         0.0           5.0  2022-03-11  20:54:27  
3               0.0         0.0           5.0  2022-02-02  16:57:59  
4               0.0         0.0           5.0  2022-02-03  17:47:44

In [96]:
df.dtypes

Orderdate             datetime64[ns]
salesValue                   float64
ItemsQty                     float64
DeliverysalesValue           float64
DeliveryitemsQty             float64
OfferSales                   float64
TimePerorder                 float64
Date                          object
Time                          object
dtype: object

In [97]:
a= time(hour = 7, minute = 0, second = 0)
b= time(hour = 14, minute = 0, second = 0)
c= time(hour = 0, minute = 0, second = 0)

df['Time_Interval'] = df['Time'].apply(lambda x: 'Dawn' if (x <= a and x > c) else ('Day' if (x <= b and x > a) else 'Night'))

In [98]:
df.head()

Orderdate  salesValue  ItemsQty  DeliverysalesValue  \
0 2022-02-24 16:23:54       122.5      4.25                 0.0   
1 2022-01-17 21:13:47        35.0      1.00                 0.0   
2 2022-03-11 20:54:27       157.5      7.00                 0.0   
3 2022-02-02 16:57:59        85.0      2.00                 0.0   
4 2022-02-03 17:47:44       100.0      1.00                 0.0   

   DeliveryitemsQty  OfferSales  TimePerorder        Date      Time  \
0               0.0         0.0           5.0  2022-02-24  16:23:54   
1               0.0         0.0           5.0  2022-01-17  21:13:47   
2               0.0         0.0           5.0  2022-03-11  20:54:27   
3               0.0         0.0           5.0  2022-02-02  16:57:59   
4               0.0         0.0           5.0  2022-02-03  17:47:44   

  Time_Interval  
0         Night  
1         Night  
2         Night  
3         Night  
4         Night

In [99]:
a= df['Time_Interval'].value_counts()
df['Time_Interval'].value_counts(normalize= True)*100

Night    75.520498
Dawn     12.452049
Day      12.027453
Name: Time_Interval, dtype: float64

In [145]:
obs1=px.bar(data_frame=a,x=a.index,y=a.values,color=a.index,template="plotly_dark"
            , labels={'y': 'Total Counts','index':"Time_Interval"},title="The Effect of Time (Dawn/Day/Night) on Sales Count")
obs1

75.52% of orders are done between 2pm and 12pm

12.45% of orders between 12am and 7am

12.03% of orders between 7am and 2pm

In [101]:
df['Date'] = pd.to_datetime(df['Date'])

In [102]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 829494 entries, 0 to 829493
Data columns (total 10 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   Orderdate           829494 non-null  datetime64[ns]
 1   salesValue          829494 non-null  float64       
 2   ItemsQty            829494 non-null  float64       
 3   DeliverysalesValue  829494 non-null  float64       
 4   DeliveryitemsQty    829494 non-null  float64       
 5   OfferSales          829494 non-null  float64       
 6   TimePerorder        829494 non-null  float64       
 7   Date                829494 non-null  datetime64[ns]
 8   Time                829494 non-null  object        
 9   Time_Interval       829494 non-null  object        
dtypes: datetime64[ns](2), float64(6), object(2)
memory usage: 63.3+ MB


In [103]:
sales_per_day = df.groupby('Date')['salesValue'].sum().reset_index()

In [126]:
sales_per_day = sales_per_day.set_index('Date')
sales_per_day.index

DatetimeIndex(['2022-01-01', '2022-01-02', '2022-01-03', '2022-01-04',
               '2022-01-05', '2022-01-06', '2022-01-07', '2022-01-08',
               '2022-01-09', '2022-01-10',
               ...
               '2022-05-13', '2022-05-14', '2022-05-15', '2022-05-16',
               '2022-05-17', '2022-05-18', '2022-05-19', '2022-05-20',
               '2022-05-21', '2022-05-22'],
              dtype='datetime64[ns]', name='Date', length=142, freq=None)

In [105]:
y = sales_per_day['salesValue'].resample('MS').mean()

In [106]:
px.line(y ,template="plotly_dark")

There is a significant increase in the mean sales from March to April

`100*(y['2022-04-01'] - y['2022-03-01'])/y['2022-03-01']`

~ 97.9% increase 

can't judge May because the data is up to 2022-05-22

In [144]:
plot = go.Figure(data=[go.Scatter(
	y=sales_per_day.values,
	mode='lines',)
])


plot.update_layout(
	updatemenus=[
		dict(
			buttons=list([
				dict(
					args=[{'y':[sales_per_day['salesValue'].resample('MS').mean()],
            'x' :[sales_per_day['salesValue'].resample('MS').mean().index]}],
					label="Sales Value Over the Month",
					method="restyle"
				),
				dict(
					args=[{'y':[sales_per_day['salesValue'].resample('W').mean()],
            'x' :[sales_per_day['salesValue'].resample('W').mean().index]}],
					label="Sales Value Over the Week",
					method="restyle"
				),
      dict(
					args=[{'y':[sales_per_day['salesValue'].resample('SM').mean()],
            'x' :[sales_per_day['salesValue'].resample('SM').mean().index]}],
					label="Sales Value Over the Semi-Month (15 Days)",
					method="restyle"
				),
      dict(
					args=[{'y':[sales_per_day['salesValue'].resample('D').mean()],
            'x' :[sales_per_day['salesValue'].resample('D').mean().index]}],
					label="Sales Value Over the Day",
					method="restyle"
				)
			]),
			direction="down",
		),
	]
)

plot.show()
